File: sdk/go/bundlebasesdk/function_serve.go
package bundlebasesdk

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sync"
	"time"

	"github.com/apache/arrow-go/v18/arrow"
	"github.com/apache/arrow-go/v18/arrow/array"
	"github.com/apache/arrow-go/v18/arrow/memory"
)

// ServeFunction runs the function provider as a JSON-RPC subprocess on stdin/stdout.
//
// If the first command-line argument is --bundlebase-functions,
// prints the function manifest as JSON and exits.
func ServeFunction(provider FunctionProvider) {
	if len(os.Args) > 1 && os.Args[1] == "--bundlebase-functions" {
		manifest := provider.Metadata()
		data, err := json.Marshal(manifest)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Failed to serialize manifest: %v\n", err)
			os.Exit(1)
		}
		os.Stdout.Write(data)
		os.Stdout.Write([]byte("\n"))
		return
	}
	ServeFunctionIO(provider, os.Stdin, os.Stdout)
}

// ServeFunctionIO runs the function provider on the given reader/writer (for testing).
func ServeFunctionIO(provider FunctionProvider, r io.Reader, w io.Writer) {
	br := bufio.NewReaderSize(r, 16*1024*1024)
	store := &stateStore{
		states:     make(map[string]interface{}),
		lastAccess: make(map[string]time.Time),
	}
	lastCleanup := time.Now()

	for {
		// Periodically clean up expired aggregate state
		if now := time.Now(); now.Sub(lastCleanup) >= 60*time.Second {
			store.cleanup(5 * time.Minute)
			lastCleanup = now
		}

		line, err := br.ReadBytes('\n')
		if len(line) > 0 {
			// Trim the trailing newline
			trimmed := line
			if trimmed[len(trimmed)-1] == '\n' {
				trimmed = trimmed[:len(trimmed)-1]
			}
			if len(trimmed) == 0 {
				continue
			}

			var req jsonRpcRequest
			if jsonErr := json.Unmarshal(trimmed, &req); jsonErr != nil {
				writeError(w, nil, -32700, fmt.Sprintf("Parse error: %v", jsonErr))
				continue
			}

			shouldStop := handleFunctionRequest(provider, store, req, br, w)
			if shouldStop {
				return
			}
		}
		if err != nil {
			return
		}
	}
}

type stateStore struct {
	mu         sync.Mutex
	states     map[string]interface{}
	lastAccess map[string]time.Time
	nextID     uint64
}

func (s *stateStore) add(state interface{}) string {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.nextID++
	id := fmt.Sprintf("state_%d", s.nextID)
	s.states[id] = state
	if s.lastAccess == nil {
		s.lastAccess = make(map[string]time.Time)
	}
	s.lastAccess[id] = time.Now()
	return id
}

func (s *stateStore) get(id string) (interface{}, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	v, ok := s.states[id]
	if ok {
		// Update last-access time for TTL
		s.lastAccess[id] = time.Now()
	}
	return v, ok
}

func (s *stateStore) set(id string, state interface{}) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.states[id] = state
	// Update last-access time for TTL
	s.lastAccess[id] = time.Now()
}

func (s *stateStore) remove(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.states, id)
	delete(s.lastAccess, id)
}

func (s *stateStore) cleanup(ttl time.Duration) {
	s.mu.Lock()
	defer s.mu.Unlock()
	now := time.Now()
	for id, accessed := range s.lastAccess {
		if now.Sub(accessed) > ttl {
			delete(s.states, id)
			delete(s.lastAccess, id)
		}
	}
}

func handleFunctionRequest(provider FunctionProvider, store *stateStore, req jsonRpcRequest, r io.Reader, w io.Writer) bool {
	switch req.Method {
	case "handshake":
		writeResponse(w, req.ID, map[string]string{"protocol_version": "1"})
	case "ping":
		writeResponse(w, req.ID, "pong")
	case "manifest":
		manifest := provider.Metadata()
		writeResponse(w, req.ID, manifest)
	case "invoke":
		handleInvoke(provider, req, r, w)
	case "create_state":
		handleCreateState(provider, store, req, w)
	case "accumulate":
		handleAccumulate(provider, store, req, r, w)
	case "merge":
		handleMerge(provider, store, req, w)
	case "evaluate":
		handleEvaluate(provider, store, req, w)
	case "shutdown":
		writeResponse(w, req.ID, map[string]bool{"ok": true})
		return true
	default:
		writeError(w, req.ID, -32601, fmt.Sprintf("Method not found: %s", req.Method))
	}
	return false
}

func handleInvoke(provider FunctionProvider, req jsonRpcRequest, r io.Reader, w io.Writer) {
	funcName, _ := req.Params["function"].(string)
	functions := provider.Functions()
	ref, ok := functions[funcName]
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function not found: %s", funcName))
		return
	}

	if ref.Scalar == nil {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function '%s' is not a scalar function", funcName))
		return
	}

	// Send ack before reading Arrow IPC (host blocks on ack before sending data)
	writeResponse(w, req.ID, map[string]bool{"ok": true})

	// Read Arrow IPC input
	records, err := readArrowIPC(r)
	if err != nil {
		// After ack, errors must be written as empty Arrow IPC (host is reading IPC, not JSON)
		var emptyBuf bytes.Buffer
		writeArrowIPC(&emptyBuf, nil)
		w.Write(emptyBuf.Bytes())
		return
	}

	// Pass the first record batch directly to the function
	var inputBatch arrow.Record
	if len(records) > 0 {
		inputBatch = records[0]
	}

	result, err := ref.Scalar.Invoke(inputBatch)
	if err != nil {
		var emptyBuf bytes.Buffer
		writeArrowIPC(&emptyBuf, nil)
		w.Write(emptyBuf.Bytes())
		return
	}

	// Build a single-column record batch from the result
	field := arrow.Field{Name: "result", Type: result.DataType()}
	schema := arrow.NewSchema([]arrow.Field{field}, nil)
	rec := array.NewRecord(schema, []arrow.Array{result}, int64(result.Len()))

	// Write Arrow IPC output
	var arrowBuf bytes.Buffer
	if err := writeArrowIPC(&arrowBuf, []arrow.Record{rec}); err != nil {
		return
	}

	w.Write(arrowBuf.Bytes())
}

func handleCreateState(provider FunctionProvider, store *stateStore, req jsonRpcRequest, w io.Writer) {
	funcName, _ := req.Params["function"].(string)
	functions := provider.Functions()
	ref, ok := functions[funcName]
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function not found: %s", funcName))
		return
	}

	if ref.Aggregate == nil {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function '%s' is not an aggregate function", funcName))
		return
	}

	state, err := ref.Aggregate.CreateState()
	if err != nil {
		writeError(w, req.ID, -32000, err.Error())
		return
	}

	id := store.add(state)
	writeResponse(w, req.ID, map[string]string{"state_id": id})
}

func handleAccumulate(provider FunctionProvider, store *stateStore, req jsonRpcRequest, r io.Reader, w io.Writer) {
	funcName, _ := req.Params["function"].(string)
	stateID, _ := req.Params["state_id"].(string)

	functions := provider.Functions()
	ref, ok := functions[funcName]
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function not found: %s", funcName))
		return
	}

	if ref.Aggregate == nil {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function '%s' is not an aggregate function", funcName))
		return
	}

	state, ok := store.get(stateID)
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("State not found: %s", stateID))
		return
	}

	// Send ack before reading Arrow IPC (host blocks on ack before sending data)
	writeResponse(w, req.ID, map[string]bool{"ok": true})

	// Read Arrow IPC input
	records, err := readArrowIPC(r)
	if err != nil {
		return
	}

	// Pass the first record batch directly to the function
	var inputBatch arrow.Record
	if len(records) > 0 {
		inputBatch = records[0]
	}

	newState, err := ref.Aggregate.Accumulate(state, inputBatch)
	if err != nil {
		return
	}

	store.set(stateID, newState)
}

func handleMerge(provider FunctionProvider, store *stateStore, req jsonRpcRequest, w io.Writer) {
	funcName, _ := req.Params["function"].(string)
	stateIDA, _ := req.Params["state_id1"].(string)
	stateIDB, _ := req.Params["state_id2"].(string)

	functions := provider.Functions()
	ref, ok := functions[funcName]
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function not found: %s", funcName))
		return
	}

	if ref.Aggregate == nil {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function '%s' is not an aggregate function", funcName))
		return
	}

	stateA, ok := store.get(stateIDA)
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("State not found: %s", stateIDA))
		return
	}

	stateB, ok := store.get(stateIDB)
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("State not found: %s", stateIDB))
		return
	}

	merged, err := ref.Aggregate.Merge(stateA, stateB)
	if err != nil {
		writeError(w, req.ID, -32000, err.Error())
		return
	}

	store.set(stateIDA, merged)
	store.remove(stateIDB)
	writeResponse(w, req.ID, map[string]string{"state_id": stateIDA})
}

func handleEvaluate(provider FunctionProvider, store *stateStore, req jsonRpcRequest, w io.Writer) {
	funcName, _ := req.Params["function"].(string)
	stateID, _ := req.Params["state_id"].(string)

	functions := provider.Functions()
	ref, ok := functions[funcName]
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function not found: %s", funcName))
		return
	}

	if ref.Aggregate == nil {
		writeError(w, req.ID, -32000, fmt.Sprintf("Function '%s' is not an aggregate function", funcName))
		return
	}

	state, ok := store.get(stateID)
	if !ok {
		writeError(w, req.ID, -32000, fmt.Sprintf("State not found: %s", stateID))
		return
	}

	result, err := ref.Aggregate.Evaluate(state)
	if err != nil {
		writeError(w, req.ID, -32000, err.Error())
		return
	}

	// Convert the result to an Arrow array and write as IPC
	alloc := memory.NewGoAllocator()
	var resultArr arrow.Array

	switch v := result.(type) {
	case int64:
		b := array.NewInt64Builder(alloc)
		b.Append(v)
		resultArr = b.NewArray()
	case float64:
		b := array.NewFloat64Builder(alloc)
		b.Append(v)
		resultArr = b.NewArray()
	case string:
		b := array.NewStringBuilder(alloc)
		b.Append(v)
		resultArr = b.NewArray()
	case bool:
		b := array.NewBooleanBuilder(alloc)
		b.Append(v)
		resultArr = b.NewArray()
	default:
		writeError(w, req.ID, -32000, fmt.Sprintf("Unsupported evaluate result type: %T", result))
		return
	}

	field := arrow.Field{Name: "result", Type: resultArr.DataType()}
	schema := arrow.NewSchema([]arrow.Field{field}, nil)
	rec := array.NewRecord(schema, []arrow.Array{resultArr}, 1)

	// Buffer Arrow IPC before sending ack
	var arrowBuf bytes.Buffer
	if err := writeArrowIPC(&arrowBuf, []arrow.Record{rec}); err != nil {
		writeError(w, req.ID, -32000, fmt.Sprintf("failed to serialize Arrow IPC result: %v", err))
		return
	}

	writeResponse(w, req.ID, map[string]bool{"ok": true})
	w.Write(arrowBuf.Bytes())

	store.remove(stateID)
}

File: sdk/java/src/main/java/com/bundlebase/sdk/FunctionServe.java
package com.bundlebase.sdk;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.*;
import org.apache.arrow.vector.ipc.ArrowStreamReader;
import org.apache.arrow.vector.ipc.ArrowStreamWriter;
import org.apache.arrow.vector.types.pojo.ArrowType;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.Schema;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.channels.Channels;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Entry point for running a function provider as a JSON-RPC subprocess.
 */
public class FunctionServe {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final BufferAllocator ALLOCATOR = new RootAllocator();

    /** State TTL: 300 seconds. */
    private static final long STATE_TTL_MS = 300_000;
    /** Cleanup interval: 60 seconds. */
    private static final long CLEANUP_INTERVAL_MS = 60_000;

    /**
     * Run the function provider on stdin/stdout.
     *
     * <p>If the first command-line argument is {@code --bundlebase-functions},
     * prints the function manifest as JSON and exits.
     */
    public static void run(Function.FunctionProvider provider, String[] args) {
        if (args != null && args.length > 0 && "--bundlebase-functions".equals(args[0])) {
            try {
                String json = MAPPER.writeValueAsString(provider.metadata());
                System.out.println(json);
            } catch (Exception e) {
                System.err.println("Failed to serialize manifest: " + e.getMessage());
                System.exit(1);
            }
            return;
        }
        run(provider, System.in, System.out);
    }

    /**
     * Run the function provider on the given streams (for testing).
     */
    public static void run(Function.FunctionProvider provider, InputStream in, OutputStream out) {
        BufferedReader reader = new BufferedReader(new InputStreamReader(in));
        StateStore store = new StateStore();
        long lastCleanup = System.currentTimeMillis();

        try {
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.isBlank()) continue;

                // Periodic cleanup of expired aggregate state
                long now = System.currentTimeMillis();
                if (now - lastCleanup >= CLEANUP_INTERVAL_MS) {
                    store.cleanup(STATE_TTL_MS);
                    lastCleanup = now;
                }

                JsonNode req;
                try {
                    req = MAPPER.readTree(line);
                } catch (Exception e) {
                    Protocol.writeError(out, null, -32700, "Parse error: " + e.getMessage());
                    continue;
                }

                String method = req.has("method") ? req.get("method").asText() : "";
                JsonNode id = req.get("id");
                JsonNode params = req.has("params") ? req.get("params") : null;

                try {
                    boolean shouldStop = handleRequest(provider, store, method, id, params, in, out);
                    if (shouldStop) {
                        break;
                    }
                } catch (Exception e) {
                    Protocol.writeError(out, id, -32000, e.getMessage());
                }
            }
        } catch (IOException e) {
            System.err.println("Error reading from stdin: " + e.getMessage());
        }
    }

    private static boolean handleRequest(
            Function.FunctionProvider provider, StateStore store,
            String method, JsonNode id, JsonNode params,
            InputStream in, OutputStream out) throws IOException {

        switch (method) {
            case "handshake" -> Protocol.writeResponse(out, id, Map.of("protocol_version", "1"));
            case "ping" -> Protocol.writeResponse(out, id, "pong");
            case "manifest" -> Protocol.writeResponse(out, id, provider.metadata());
            case "invoke" -> handleInvoke(provider, id, params, in, out);
            case "create_state" -> handleCreateState(provider, store, id, params, out);
            case "accumulate" -> handleAccumulate(provider, store, id, params, in, out);
            case "merge" -> handleMerge(provider, store, id, params, out);
            case "evaluate" -> handleEvaluate(provider, store, id, params, out);
            case "shutdown" -> {
                Protocol.writeResponse(out, id, Map.of("ok", true));
                return true;
            }
            default -> Protocol.writeError(out, id, -32601, "Method not found: " + method);
        }
        return false;
    }

    private static void handleInvoke(
            Function.FunctionProvider provider, JsonNode id, JsonNode params,
            InputStream in, OutputStream out) throws IOException {
        String funcName = params != null && params.has("function") ? params.get("function").asText() : "";
        Object fn = provider.functions().get(funcName);
        if (fn == null) {
            Protocol.writeError(out, id, -32000, "Function not found: " + funcName);
            return;
        }
        if (!(fn instanceof Function.ScalarFunction scalar)) {
            Protocol.writeError(out, id, -32000,
                    "Function '" + funcName + "' is not a scalar function (actual type: " + fn.getClass().getName() + ")");
            return;
        }

        // Read Arrow IPC input
        VectorSchemaRoot inputRoot = readArrowIPC(in);
        List<FieldVector> args = new ArrayList<>();
        if (inputRoot != null) {
            args.addAll(inputRoot.getFieldVectors());
        }

        FieldVector result;
        try {
            result = scalar.invoke(args);
        } catch (Exception e) {
            Protocol.writeError(out, id, -32000, e.getMessage());
            if (inputRoot != null) inputRoot.close();
            return;
        }

        // Build a single-column VectorSchemaRoot from the result
        Schema schema = new Schema(List.of(result.getField()));
        VectorSchemaRoot resultRoot = new VectorSchemaRoot(List.of(result.getField()), List.of(result), result.getValueCount());

        // Buffer Arrow IPC before sending ack
        ByteArrayOutputStream arrowBuf = new ByteArrayOutputStream();
        try {
            Protocol.writeArrowIPC(arrowBuf, resultRoot);
        } catch (IOException e) {
            Protocol.writeError(out, id, -32000, "Failed to serialize Arrow IPC result: " + e.getMessage());
            resultRoot.close();
            if (inputRoot != null) inputRoot.close();
            return;
        }

        Protocol.writeResponse(out, id, Map.of("ok", true));
        out.write(arrowBuf.toByteArray());
        out.flush();

        if (inputRoot != null) inputRoot.close();
    }

    @SuppressWarnings("unchecked")
    private static void handleCreateState(
            Function.FunctionProvider provider, StateStore store,
            JsonNode id, JsonNode params, OutputStream out) throws IOException {
        String funcName = params != null && params.has("function") ? params.get("function").asText() : "";
        Object fn = provider.functions().get(funcName);
        if (fn == null) {
            Protocol.writeError(out, id, -32000, "Function not found: " + funcName);
            return;
        }
        if (!(fn instanceof Function.AggregateFunction<?> agg)) {
            Protocol.writeError(out, id, -32000,
                    "Function '" + funcName + "' is not an aggregate function (actual type: " + fn.getClass().getName() + ")");
            return;
        }

        Object state;
        try {
            state = agg.createState();
        } catch (Exception e) {
            Protocol.writeError(out, id, -32000, e.getMessage());
            return;
        }

        String stateId = store.add(state);
        Protocol.writeResponse(out, id, Map.of("state_id", stateId));
    }

    @SuppressWarnings("unchecked")
    private static void handleAccumulate(
            Function.FunctionProvider provider, StateStore store,
            JsonNode id, JsonNode params, InputStream in, OutputStream out) throws IOException {
        String funcName = params != null && params.has("function") ? params.get("function").asText() : "";
        String stateId = params != null && params.has("state_id") ? params.get("state_id").asText() : "";

        Object fn = provider.functions().get(funcName);
        if (fn == null) {
            Protocol.writeError(out, id, -32000, "Function not found: " + funcName);
            return;
        }
        if (!(fn instanceof Function.AggregateFunction agg)) {
            Protocol.writeError(out, id, -32000,
                    "Function '" + funcName + "' is not an aggregate function (actual type: " + fn.getClass().getName() + ")");
            return;
        }

        Object state = store.get(stateId);
        if (state == null) {
            Protocol.writeError(out, id, -32000, "State not found: " + stateId);
            return;
        }

        // Read Arrow IPC input
        VectorSchemaRoot inputRoot = readArrowIPC(in);
        List<FieldVector> args = new ArrayList<>();
        if (inputRoot != null) {
            args.addAll(inputRoot.getFieldVectors());
        }

        Object newState;
        try {
            newState = agg.accumulate(state, args);
        } catch (Exception e) {
            Protocol.writeError(out, id, -32000, e.getMessage());
            if (inputRoot != null) inputRoot.close();
            return;
        }

        store.set(stateId, newState);
        Protocol.writeResponse(out, id, Map.of("ok", true));
        if (inputRoot != null) inputRoot.close();
    }

    @SuppressWarnings("unchecked")
    private static void handleMerge(
            Function.FunctionProvider provider, StateStore store,
            JsonNode id, JsonNode params, OutputStream out) throws IOException {
        String funcName = params != null && params.has("function") ? params.get("function").asText() : "";
        String stateIdA = params != null && params.has("state_id_a") ? params.get("state_id_a").asText() : "";
        String stateIdB = params != null && params.has("state_id_b") ? params.get("state_id_b").asText() : "";

        Object fn = provider.functions().get(funcName);
        if (fn == null) {
            Protocol.writeError(out, id, -32000, "Function not found: " + funcName);
            return;
        }
        if (!(fn instanceof Function.AggregateFunction agg)) {
            Protocol.writeError(out, id, -32000,
                    "Function '" + funcName + "' is not an aggregate function (actual type: " + fn.getClass().getName() + ")");
            return;
        }

        Object stateA = store.get(stateIdA);
        if (stateA == null) {
            Protocol.writeError(out, id, -32000, "State not found: " + stateIdA);
            return;
        }
        Object stateB = store.get(stateIdB);
        if (stateB == null) {
            Protocol.writeError(out, id, -32000, "State not found: " + stateIdB);
            return;
        }

        Object merged;
        try {
            merged = agg.merge(stateA, stateB);
        } catch (Exception e) {
            Protocol.writeError(out, id, -32000, e.getMessage());
            return;
        }

        store.set(stateIdA, merged);
        store.remove(stateIdB);
        Protocol.writeResponse(out, id, Map.of("ok", true));
    }

    @SuppressWarnings("unchecked")
    private static void handleEvaluate(
            Function.FunctionProvider provider, StateStore store,
            JsonNode id, JsonNode params, OutputStream out) throws IOException {
        String funcName = params != null && params.has("function") ? params.get("function").asText() : "";
        String stateId = params != null && params.has("state_id") ? params.get("state_id").asText() : "";

        Object fn = provider.functions().get(funcName);
        if (fn == null) {
            Protocol.writeError(out, id, -32000, "Function not found: " + funcName);
            return;
        }
        if (!(fn instanceof Function.AggregateFunction agg)) {
            Protocol.writeError(out, id, -32000,
                    "Function '" + funcName + "' is not an aggregate function (actual type: " + fn.getClass().getName() + ")");
            return;
        }

        Object state = store.get(stateId);
        if (state == null) {
            Protocol.writeError(out, id, -32000, "State not found: " + stateId);
            return;
        }

        Object result;
        try {
            result = agg.evaluate(state);
        } catch (Exception e) {
            Protocol.writeError(out, id, -32000, e.getMessage());
            return;
        }

        // Convert the scalar result to a single-element Arrow vector
        VectorSchemaRoot resultRoot = scalarToRoot(result);
        if (resultRoot == null) {
            Protocol.writeError(out, id, -32000, "Unsupported evaluate result type: " + result.getClass().getName());
            return;
        }

        // Buffer Arrow IPC before sending ack
        ByteArrayOutputStream arrowBuf = new ByteArrayOutputStream();
        try {
            Protocol.writeArrowIPC(arrowBuf, resultRoot);
        } catch (IOException e) {
            Protocol.writeError(out, id, -32000, "Failed to serialize Arrow IPC result: " + e.getMessage());
            resultRoot.close();
            return;
        }

        Protocol.writeResponse(out, id, Map.of("ok", true));
        out.write(arrowBuf.toByteArray());
        out.flush();

        resultRoot.close();
        store.remove(stateId);
    }

    /**
     * Read a length-prefixed Arrow IPC stream from the input.
     */
    private static VectorSchemaRoot readArrowIPC(InputStream in) throws IOException {
        byte[] lengthBuf = new byte[4];
        int read = in.read(lengthBuf);
        if (read < 4) {
            return null;
        }
        int length = ByteBuffer.wrap(lengthBuf).getInt();
        if (length == 0) {
            return null;
        }

        byte[] data = in.readNBytes(length);
        if (data.length < length) {
            throw new IOException("Unexpected end of Arrow IPC data");
        }

        BufferAllocator allocator = new RootAllocator();
        ArrowStreamReader reader = new ArrowStreamReader(new ByteArrayInputStream(data), allocator);
        if (!reader.loadNextBatch()) {
            reader.close();
            return null;
        }

        // Transfer ownership of vectors so we can close the reader
        VectorSchemaRoot root = reader.getVectorSchemaRoot();
        // We need to keep the root alive, so we don't close the reader here.
        // The caller is responsible for closing the root.
        return root;
    }

    /**
     * Convert a scalar value to a single-row VectorSchemaRoot.
     */
    private static VectorSchemaRoot scalarToRoot(Object value) {
        BufferAllocator allocator = new RootAllocator();

        if (value instanceof Long l) {
            Schema schema = new Schema(List.of(Field.nullable("result", new ArrowType.Int(64, true))));
            VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator);
            root.allocateNew();
            ((BigIntVector) root.getVector("result")).setSafe(0, l);
            root.setRowCount(1);
            return root;
        } else if (value instanceof Integer i) {
            Schema schema = new Schema(List.of(Field.nullable("result", new ArrowType.Int(64, true))));
            VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator);
            root.allocateNew();
            ((BigIntVector) root.getVector("result")).setSafe(0, i.longValue());
            root.setRowCount(1);
            return root;
        } else if (value instanceof Double d) {
            Schema schema = new Schema(List.of(Field.nullable("result", new ArrowType.FloatingPoint(
                    org.apache.arrow.vector.types.FloatingPointPrecision.DOUBLE))));
            VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator);
            root.allocateNew();
            ((Float8Vector) root.getVector("result")).setSafe(0, d);
            root.setRowCount(1);
            return root;
        } else if (value instanceof String s) {
            Schema schema = new Schema(List.of(Field.nullable("result", new ArrowType.Utf8())));
            VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator);
            root.allocateNew();
            ((VarCharVector) root.getVector("result")).setSafe(0, s.getBytes());
            root.setRowCount(1);
            return root;
        } else if (value instanceof Boolean b) {
            Schema schema = new Schema(List.of(Field.nullable("result", new ArrowType.Bool())));
            VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator);
            root.allocateNew();
            ((BitVector) root.getVector("result")).setSafe(0, b ? 1 : 0);
            root.setRowCount(1);
            return root;
        }

        return null;
    }

    /**
     * Thread-safe state store for aggregate function accumulators.
     */
    static class StateStore {
        private final ConcurrentHashMap<String, Object> states = new ConcurrentHashMap<>();
        private final ConcurrentHashMap<String, Long> createdAt = new ConcurrentHashMap<>();
        private final AtomicLong nextId = new AtomicLong(0);

        String add(Object state) {
            long id = nextId.incrementAndGet();
            String key = "state_" + id;
            states.put(key, state);
            createdAt.put(key, System.currentTimeMillis());
            return key;
        }

        Object get(String id) {
            return states.get(id);
        }

        void set(String id, Object state) {
            states.put(id, state);
        }

        void remove(String id) {
            states.remove(id);
            createdAt.remove(id);
        }

        void cleanup(long ttlMs) {
            long now = System.currentTimeMillis();
            for (Map.Entry<String, Long> entry : createdAt.entrySet()) {
                if (now - entry.getValue() > ttlMs) {
                    states.remove(entry.getKey());
                    createdAt.remove(entry.getKey());
                }
            }
        }
    }
}
